### DISCUSSION TITLE
[Ethers.js x Celo] What issues are you facing with Ethers.js?

### DISCUSSION THREAD

**Discussion by arthurgousset** [celo-org/celo-monorepo#9922] [Q&A]
👉 **TLDR: Please share/upvote issues you have with Ethers.js on Celo.** 👈 

Hi all, @aaronmgdr is investigating issues you face with [Ethers.js](https://www.npmjs.com/package/ethers). 

To help you, we'd like learn about your issues and to get as much context as possible.

Here are some problems on our radar: 

1. re **gas estimation** (missing `gasLimit` field in RPC API) has been fixed by @piersy in this PR: https://github.com/celo-org/celo-blockchain/pull/1957
2. re **paying gas with custom tokens** (missing `feeCurrency` support in Ethers.js). This is on @aaronmgdr's radar 📡

Here is some feedback we heard: 

> For me (as a developer), the only things I care about is to connect/disconnect to a wallet, get the user address, signer and provider (so I know if on the wrong network) and be able to set fee currency. [...] I don't want anything else other than a generic wallet connector and be able to set fee transactions. @celo/connect is working well for setting fees.

Our ask: 👉 **Please share/upvote issues you have with Ethers.js on Celo.** 👈

**Comment by aaronmgdr** [comment]
Also would like to know if the [celo-ethers-wrapper](https://www.npmjs.com/package/@celo-tools/celo-ethers-wrapper) has been tried by people and if so what limitations have come up. I suppose the main one is that it so far only implements the `Wallet` Signer which requires having a private key.

**Reply by d-mooers** [reply]
A very early version of Centro used `celo-ethers-wrapper`.  Back then I think the main issue was getting wallet connection working correctly with Valora (spurred the creation of [valora-ethers](https://www.npmjs.com/package/@open-celo/valora-ethers)) as contract kit and the valora connection lib used web3js.  

I havent revisited `celo-ethers-wrapper` since then, but might now

**Reply by vibern0** [reply]
A few days ago I was watching Richard during Devcon talking about ethers v6. There will be network plugins https://youtu.be/qli5SoAKC1s?t=757 If I understood correctly how it works, it might be a good chance to build a plugin and drop this wrapper.
And answering the question, I never used. What I use thought is a combination of react-celo, @celo/connect, and ethers, just so I handle everything with ethers but submit the transaction with @celo/connect so I can set fee currency, using react-celo provider.

**Comment by d-mooers** [comment]
From Node's side, the biggest blocker/request is really just being able to pay gas with a custom token and correctly estimate gas for the custom gas token.  Besides this we have not encountered many issues with ethers on Celo.

**Reply by arthurgousset** [reply]
Thanks for sharing!

**Comment by kosecki123** [comment]
> 1. re **gas estimation** (missing `gasLimit` field in RPC API) has been fixed by @piersy in this PR: [Improve RPC API compatibility for Ethers.js celo-blockchain#1957](https://github.com/celo-org/celo-blockchain/pull/1957)

Regarding this point, does this has to be deployed by nodes providers? 

Currently using vanilla JsonRpcProvider from ethers 5.7.1 with forno alfajores are giving me

```
{
  reason: 'invalid BigNumber value',
  code: 'INVALID_ARGUMENT',
  argument: 'value',
  value: undefined,
  checkKey: 'gasLimit',
  checkValue: undefined
}
```

error

**Reply by kosecki123** [reply]
I've created the repo with 2 issues documenting 2 different behaviours

1. The issue related to `gasLimit` posted here https://github.com/neutral-protocol/celo-ethers/issues/2
2. The other one that have issues with local node running from Hardhat and using CeloProvider can be found here https://github.com/neutral-protocol/celo-ethers/issues/1

**Reply by arthurgousset** [reply]
Thanks for the detailed info and re-posting your issue here @kosecki123 💯 
@aaronmgdr is currently investigating the Ethersjs incompatibilities. 
We'll take a look to see if we can reproduce the error 👍

**Reply by aaronmgdr** [reply]
> Regarding this point, does this has to be deployed by nodes providers?

yes the release it is part of is still being tested. so its not out yet. and once its out full node providers will need to run it

**Reply by kosecki123** [reply]
> yes the release it is part of is still being tested. so its not out yet. and once its out full node providers will need to run it

OK, clear. I assume to will fix the issue of using vanilla ethers provider with Celo nodes.

**Reply by aaronmgdr** [reply]
> > yes the release it is part of is still being tested. so its not out yet. and once its out full node providers will need to run it
> 
> OK, clear. I assume to will fix the issue of using vanilla ethers provider with Celo nodes.

That's right.

**Reply by kosecki123** [reply]
> That's right.

And, obvious follow up q, wen?

**Reply by aaronmgdr** [reply]
I believe the eta is about a month right now. 

@piersy since he wrote the thing

**Reply by arthurgousset** [reply]
Here is what @piersy shared 2 weeks ago: 

> The fix is in this release https://github.com/celo-org/celo-blockchain/releases/tag/v1.7.0-beta.1 but that is a beta that is currently undergoing testing on the baklava network, once done it will be promoted to alfajores and finally to mainnet. The schedule used for that is laid out [here](https://docs.celo.org/community/release-process/blockchain-client#distribution) but we have encountered a problem with baklava so the process is currently stalled while we look into that.

**Comment by aaronmgdr** [comment]
BTW with the release of 1.7.0 aka the Flan hardfork  November 9, 2022 `gasLimit` is included in RPC blocks

**Comment by aaronmgdr** [comment]
Here is doc with my research / work on fee currencies. Please write here if you see any blindspots

https://docs.google.com/document/d/1tQXBq8Hr2YuKWNXctmeCPzKvatpCBRPGSrAUtqttfog/edit?usp=sharing

**Comment by arthurgousset** [comment]
Sharing a discussion going on in the Etherjs repo re custom transaction types: 
- https://github.com/ethers-io/ethers.js/issues/3747

**Comment by amytsang** [comment]
Saw the following error message when connecting to ethersjs v6.3.0 using the RPC at https://forno.celo.org

<img width="973" alt="Screenshot 2023-05-05 at 9 39 56 AM" src="https://user-images.githubusercontent.com/6961602/236516884-5b8d3326-69ab-480a-9889-f10ca9fdf91b.png">

**Reply by karlb** [reply]
Celo does not include a `difficulty` value for blocks at the moment, as it is not relevant to our consensus. However, to increase the compatibility with Ethereum, there is a CIP draft to add this and other fields to the block header with constant values. See https://github.com/celo-org/celo-proposals/discussions/370

Without looking into the Ethers code, I think the CIP might resolve this specific issue.

**Reply by arthurgousset** [reply]
Hi @amytsang, briefly following up here:

TLDR:

+ we deployed a fix to the Forno RPC you are using at https://forno.celo.org/ 
+ you should not see the error above anymore. Could you confirm that's the case?

More context:

+ the quick fix adds a `difficulty` value to the response of the JSON-RPC API in Forno (see this [PR](https://github.com/celo-org/celo-blockchain/pull/2082))
+ Why? As @karlb mentioned, Celo block headers don't currently include a `difficulty` value. There is a [proposal](https://github.com/celo-org/celo-proposals/discussions/370) to add them back in the next hardfork, but in the mean time we artificially added them back in the response from the JSON-RPC API in Forno.
+ This means that the above request will not throw an error when using the Forno RPC, but might still throw an error in other RPC providers until the next hardfork

For future reference, the code snippet that caused this error was:

```js
import { ethers } from "ethers";

(async () => {
    const provider = new ethers.JsonRpcProvider("https://forno.celo.org");
    const feeData = await provider.getFeeData();  // <--------- BUG
})();
```

**Reply by amytsang** [reply]
Thank you, confirmed that's resolved!
